### PR TITLE
Init autoinstall configuration

### PR DESCRIPTION
**Pull request by tai271828**
## Types of changes

- **New feature**
- **Documentation Update**

## Description

Inspired by [this discourse post "Please test autoinstalls for 20.04!"](https://discourse.ubuntu.com/t/please-test-autoinstalls-for-20-04/15250), I think that both autoinstall and ubuntu-server-netboot are getting much more mature to use. This pull request initiate the implementation to integrate both of them, so we may look forward to automating our pxe testing process.

## Steps to Test This Pull Request
1. Grab a UEFI bare metal, and deploy focal to it.
2. install prerequisite package dependencies.
2. (on the deployed focal) `ubuntu-server-netboot$ sudo ./ubuntu-server-netboot.py --url http://cdimage.ubuntu.com/ubuntu/releases/20.04.2/release/ubuntu-20.04.2-live-server-arm64.iso --autoinstall-url http://<IP OF YOUR FOCAL>/`

Platform tested: awrep6 (netboot server), d05-2, scobee

## Expected Result
`ubuntu-installer/grub/grub.cfg` now has the inserted snippet ` autoinstall "ds=nocloud-net;s=http://<IP OF YOUR FOCAL>/" ` before `---`.

## Additional context

[Ubuntu Server Guide - Autoinstall Quick Start](https://ubuntu.com/server/docs/install/autoinstall-quickstart)
[Ubuntu Server Guide - Autoinstall](https://ubuntu.com/server/docs/install/autoinstall)

### CODE REVIEW

File: autoinstall/README.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# Autoinstall
+By using autoinstall we could automate the installer like what `presed` is used for `d-i`.
```
Comment (dannf) [inline autoinstall/README.md]: s/presed/preseed/

Comment (dannf) [reply autoinstall/README.md]: While at it, it might be good to spell out `debian-installer` - I'm not sure the average user would recognize the acronym. They may not recognize `debian-installer` either, but that's OK :)

File: autoinstall/README.md [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+## Usage
+1. Upload your `user-data` to a www server. You may use `basic.yaml` as an example. Rename `basic.yaml` and put it 
+   into a www server to serve it.
```
Comment (dannf) [inline autoinstall/README.md]: Do you have to rename it? If so, what should it be renamed to?

File: autoinstall/README.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+2. `touch <your www server>/meta-data` to create an empty but necessary file required by `cloud-init`.
+3. Edit `grub.cfg` to include the necessary kernel parameters `autoinstall "ds=nocloud-net;s=http://<www-server-url>/"`. For example,
```
Comment (dannf) [inline autoinstall/README.md]: `grub.cfg` will be UEFI-specific. `pxelinux` users should also be able to use it by changing their `APPEND` line in the same way, so might be good to make this generic. In fact, have you considered adding an `--autoinstall-url` option to the script to do it for you? We already have the `--extra-args` parameter which modifies the command line in both config files. Difference here is that it'd need to be inserted before the `---`

File: autoinstall/README.md [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+     - as is `linux   /casper/vmlinuz    quiet splash root=/dev/ram0 ramdisk_size=1500000 ip=dhcp url=http://<server-provides-image>/focal-live-server-arm64.iso ---`
+     - to be `linux   /casper/vmlinuz    quiet splash root=/dev/ram0 ramdisk_size=1500000 ip=dhcp url=http://<server-provides-image>/focal-live-server-arm64.iso autoinstall "ds=nocloud-net;s=http://<server-provides-user-data>/" ---`
+
+## Trouble-shooting and Tips
+- The format of the first line "shebang", a.k.a. `#cloud-config` matters. Do not change it or input extra spaces 
+  unless you know what you are doing.
+- The double quotes matters when injecting user-data in the kernel parameters of Grub2 menu.
```
Comment (dannf) [inline autoinstall/README.md]: Nit: s/matters/matter/ - and as above, it isn't GRUB-specific

File: autoinstall/basic.yaml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+#cloud-config
+autoinstall:
+  version: 1
+  identity:
+    hostname: ubuntu-server
+    password: "$6$exDY1mhS4KUYCE/2$zmn9ToZwTKLhCw.b4/b.ZRTIZM30JZ4QrOQ2aOXJ8yk96xpcCof0kxKwuX1kqLG/ygbJ1f8wxED22bTL4F46P0"
```
Comment (dannf) [inline autoinstall/basic.yaml:13]: Can you put comments here? If so, it might be good to note what actual `password` this maps to - and how to generate your own (pointing to external docs is definitely ok)

**Review by dannf — CHANGES_REQUESTED** [review]
Thanks @tai271828 ! Comments inline.

File: autoinstall/README.md (hunk 3 above)
Comment (tai271828) [reply autoinstall/README.md]: My gut feeling shows this is an good idea. Let me think more about it.

**Comment by tai271828** [comment]
Looks like a lot of improvement is suggested. Let me convert this pull request to be "draft".

File: autoinstall/README.md (hunk 2 above)
Comment (tai271828) [reply autoinstall/README.md]: Yes, `basic.yaml` should be renamed to `user-data`, which is the spec of `cloud-init`.

Comment (tai271828) [reply autoinstall/README.md]: I should have used statement like `Rename basic.yaml to be user-data` to make it clear. I will update it.

File: autoinstall/README.md (hunk 1 above)
Comment (tai271828) [reply autoinstall/README.md]: Will fix the typo and use statement like:

`preseed` is used for `debian-installer`, a.k.a. `d-i`

File: autoinstall/README.md (hunk 3 above)
Comment (tai271828) [reply autoinstall/README.md]: Using `--extra-args` sounds good, let me have a try with the script and update the README correspondingly.

File: autoinstall/README.md (hunk 4 above)
Comment (tai271828) [reply autoinstall/README.md]: Will fix. Thanks!

File: autoinstall/basic.yaml (hunk 5 above)
Comment (tai271828) [reply autoinstall/basic.yaml:13]: Good idea. Thanks for the reminder. Will do. It will look like:

```
    # The password hash maps to "ubuntu".
    # See the example of Ubuntu Server guide https://ubuntu.com/server/docs/install/autoinstall-quickstart
    #
    # You may generate your own password by:
    #   mkpasswd --method=SHA-512 --rounds=4096
    # Refer to the document of cloudinit for generating password
    # https://cloudinit.readthedocs.io/en/latest/topics/examples.html
```

File: autoinstall/README.md (hunk 3 above)
Comment (dannf) [reply autoinstall/README.md]: I don't think we want to use `--extra-args` directly - that will place the options after the `---`, which means they'll be retained on the *installed* system. We probably need something new/similar that can place the args *before* the `---`, e.g. a `--autoinstall-url` parameter

Comment (tai271828) [reply autoinstall/README.md]: Right, thanks for the reminder! I have implemented `--autoinstall-url`, and will soon update this pull request draft after making sure I have fixed according to all suggestion and discussion and rebased.

**Comment by tai271828** [comment]
I have updated this pull request and its corresponding issue description. Please note there is a instruction to show the steps for testing the new features/options.

**Review by dannf — COMMENTED** [review]
+1 on black but, if we do it, I think we need to enforce it. To that end, would you mind adding a pre-commit hook to validate it?

File: README.md [hunk 6]
```diff
@@ -19,12 +19,14 @@ You can also add additional kernel command line arguments (e.g. `"console=ttyS0"
 
 ## Usage
 Copy the files generated under the interim folder `/tmp/tmpxxx/ubuntu-installer/`
-to your tftp root folder for netboot, for example `/var/lib/tftpboot`:
+to your tftp root folder for netboot, for example `/srv/tftp` or `/var/lib/tftpboot`.
```
Comment (dannf) [inline README.md:25]: +1 - `/srv/tftp` is a more prominent example these days

File: README.md [hunk 7]
```diff
@@ -19,12 +19,14 @@ You can also add additional kernel command line arguments (e.g. `"console=ttyS0"
...
+You may check your tftpd configuration of the root directory via `/etc/default/tftpd-hpa`. Let's copy:
```
Comment (dannf) [inline README.md]: There's a lot of different TFTP server choices, and different ways to configure them on different operating systems, so I'm inclined to consider TFTP server configuration out of scope here less feature creep turn this into yet another TFTP debug guide.

I think providing examples of TFTP root folder locations is different - it's just an example, and one that's fairly standardized via the File Hierarchy Standard IIRC.

File: contrib/install-dependencies.sh [hunk 8]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+apt update
+apt install genisoimage mtools python3-distro-info -y
```
Comment (dannf) [inline contrib/install-dependencies.sh]: I'd like to make this more distro-agnostic overtime, so I'd suggest we print a warning that only Debian/Ubuntu are supported if we find we're not on one of those systems. We could then enable more OS's over time. Also seems like we could put this at the top level in e.g. a `./setup.sh` or similar instead of in a `contrib/` subdir.

File: contrib/install-dependencies.sh [hunk 9]
```diff
@@ -0,0 +1,7 @@
...
+
+# if you want this tool to setup pxe server for you
+# these packages will be useful when invoking --deploy-dir
+apt install apache2 tftpd-hpa -y
```
Comment (dannf) [inline contrib/install-dependencies.sh]: This is a little too scary to me. While nobody should be running scripts off the internet without understanding what they do - and certainly not as root as this requires - some people will, and I wouldn't want to open up listening services on their system w/o express permission.

File: ubuntu-server-netboot.py [hunk 10]
```diff
@@ -228,6 +241,10 @@ def setup_autoinstall(apache_dir="/var/www/html"):
         "--autoinstall-url",
         help="URL to Autoinstall config file to be used during Subiquity installation",
     )
+    parser.add_argument(
+        "--deploy-dir",
+        help="Deploy the stating dir to the target tftp dir. sudo permission is needed."
```
Comment (dannf) [inline ubuntu-server-netboot.py]: I'm not really sure how this differs from `--out-dir`. Also note that sudo permission isn't required - you just need permission to write to the target dir.

**Review by dannf — CHANGES_REQUESTED** [review]
Sorry, hit EOD before I could complete this review, but i have some comments in there. One thing I'd suggest is to split the `black` bit - ideally with an enforcing `pre-commit` hook into it's own pull request and we can merge that first, then dig into the rest w/ more focus.

File: README.md (hunk 7 above)
Comment (tai271828) [reply README.md]: Fair. Let me update this part.

File: contrib/install-dependencies.sh (hunk 8 above)
Comment (tai271828) [reply contrib/install-dependencies.sh]: That's great! I am thinking the same thing as well. Let me file a feature request for the implementation of dependency checker. See https://github.com/dannf/ubuntu-server-netboot/issues/12 .

Comment (tai271828) [reply contrib/install-dependencies.sh]: I prefer to implement this later on, and focus on `--autoinstall-url` instead. What do you think?

File: contrib/install-dependencies.sh (hunk 9 above)
Comment (tai271828) [reply contrib/install-dependencies.sh]: Understood. Let me strip `./contrib` and focus on `--autoinstall-url`.

**Comment by tai271828** [comment]
Sorry for unclear statement. The commit of black style fix is individual. Let me remove the commit for easier review. As what you proposed, we could merge the fix from pre-commit hook in another pull request. Trace this feature request by https://github.com/dannf/ubuntu-server-netboot/issues/13 .

File: ubuntu-server-netboot.py (hunk 10 above)
Comment (tai271828) [reply ubuntu-server-netboot.py]: `--out-dir` is used to output `ubuntu-installer` somewhere, and `--deploy-dir` is used to move files to tftp root folder (and apache server if autoinstall is enabled).


What I want to achieve is a "one line command" to setup a netboot server, so

1. I hope the tool could copy files to `/srv/tftp` directly, and it needs the directory permission (via sudo).
2. `--out-dir` has a default sub-directory `ubuntu-installer`, and `grubnetaa64.efi` does not support the `ubuntu-installer` structure now if I understand correctly.


I may be too obsessed with the concept of "one line command to deploy". The support of `--deploy-dir` is also an individual commit. Maybe I could remove it in this pull request, and let us focus on `--autoinstall-url` better.

File: contrib/install-dependencies.sh (hunk 8 above)
Comment (dannf) [reply contrib/install-dependencies.sh]: Either order works for me, thanks!

File: ubuntu-server-netboot.py (hunk 10 above)
Comment (dannf) [reply ubuntu-server-netboot.py]: I think `--autoinstall-url` would be easy to understand and useful on its own, so I'd +1 tackling that first. TBH, I'm less sure about the maintainability of dealing with the preseed/http setup ourselves. Providing example preseeds seems like a good idea, but I'm not sure how much further we can go while still doing it well.

**Comment by dannf** [comment]
> Sorry for unclear statement. The commit of black style fix is individual. Let me remove the commit for easier review. As what you proposed, we could merge the fix from pre-commit hook in another pull request. Trace this feature request by #13 .

Oh, yeah, I think the black style fix is fine, and fine to include here as you've done - I just think we should add the `.pre-commit.yaml` at the same time to make sure that your cleanup doesn't bit rot with future commits :)

File: ubuntu-server-netboot.py (hunk 10 above)
Comment (tai271828) [reply ubuntu-server-netboot.py]: +1 on dropping the maintainability of dealing with the preseed/http setup ourselves. I will update the pr later on.

**Comment by tai271828** [comment]
Re-upload the pull request with the re-based and updated code-base.

The update/fix is implemented by following our previous discussion and conclusion, including:
1. Only focusing on the implementation of `--autoinstall-url` for `grub.cfg`. Drop the other commits, features and style fixes.
2. Update the issue description

The features removed from the earlier version of this pull request are moved to [another repository](https://github.com/tai271828/ubuntu-autoinstall-pxe-server), which is essentially a command line wrapper of this tool, `ubuntu-server-netboot`. In my opinion, we could have two advantages 1) we could keep `ubuntu-server-netboot` doing one thing only and doing it well 2) satisfying the system admins who need to manually setup their own pxe server manually very often, so they could still have choice by using [another wrapper](https://github.com/tai271828/ubuntu-autoinstall-pxe-server).

File: README.md [hunk 11]
```diff
@@ -15,14 +15,18 @@ The `--url` parameter is used for 2 reasons:
...
+Optionally, you can place `--autoinstall-url` to tell the netbooting process to enable subiquity automation. See [our autoinstall example](./autoinstall/README.md) and [the autoinstall and Automated Server Installs
+Introduction of Ubuntu Server guide](Automated Server Installs Introduction) for more details.
+
 You can also add additional kernel command line arguments (e.g. `"console=ttyS0"`) to the generated configuration files using the `--extra-args` parameter.
 
-## Usage
+## Usage of the Generated Files
 Copy the files generated under the interim folder `/tmp/tmpxxx/ubuntu-installer/`
-to your tftp root folder for netboot, for example `/var/lib/tftpboot`:
+to your tftp root folder for netboot, for example `/srv/tftp` or `/var/lib/tftpboot`.
```
Comment (dannf) [inline README.md:25]: The changes in the `Usage` section seem to be orthogonal to the autoinstall feature. I agree with both of the changes, but can you make them in a separate commit?

File: README.md (hunk 7 above)
Comment (dannf) [reply README.md]: I think we should omit the line about `/etc/default/tftpd-hpa` per the argument above. `tftpd-hpa` is just one TFTP server option, and having it's configuration in `/etc/default/tftpd-hpa` is specific to Debian/Ubuntu - so not everyone can expect it to be there. There's too many possibilities for how to configure a TFTP server to try and cover them all here - better the user consult their docs.

File: README.md [hunk 12]
```diff
@@ -15,14 +15,18 @@ The `--url` parameter is used for 2 reasons:
...
+You may check your tftpd configuration of the root directory via `/etc/default/tftpd-hpa`. Let's copy:
 
 ```
-$ sudo cp /tmp/tmpxxx/ubuntu-installer/* /var/lib/tftpboot/
+$ sudo cp -r /tmp/tmpxxx/ubuntu-installer/* /srv/tftp
```
Comment (dannf) [inline README.md:29]: While I agree with this, can you split it into a separate commit (maybe combined w/ the other `Usage` changes?)

File: autoinstall/README.md [hunk 13]
```diff
@@ -0,0 +1,17 @@
+# Autoinstall
+By using autoinstall we could automate the subiquity installer like what `preseed` is used for `debian-installer`, a.k.a. `d-i`.
+
+## Usage
+1. Upload your `user-data` to a www server. You may use `basic.yaml` as an example. Rename `basic.yaml` to be `user-data` and put it into a www server to serve it.
```
Comment (dannf) [inline autoinstall/README.md]: Mind noting that there's a pre-baked password within that they may want to change?

File: ubuntu-server-netboot.py [hunk 14]
```diff
@@ -251,7 +255,12 @@ def cleanup(directory):
     )
     os.mkdir(os.path.join(staging_dir, "grub"))
     with open(os.path.join(staging_dir, "grub", "grub.cfg"), "w") as grub_cfg:
-        netboot_args = "%s url=%s" % (Netboot_Args, args.url)
+        if args.autoinstall_url:
+            autoinstall_args = f'"ds=nocloud-net;s={args.autoinstall_url}"'
+            netboot_args = "%s url=%s autoinstall %s" % (Netboot_Args, args.url, autoinstall_args)
+        else:
+            netboot_args = "%s url=%s" % (Netboot_Args, args.url)
+
```
Comment (dannf) [inline ubuntu-server-netboot.py:263]: remember to also do the same for pxelinux - autoinstall isn't specific to UEFI.

File: ubuntu-server-netboot.py [hunk 15]
```diff
@@ -297,3 +306,4 @@ def cleanup(directory):
             pxelinux_cfg.write("\n")
     atexit.unregister(cleanup)
     logger.info("Netboot generation complete: %s" % (staging_dir))
+
```
Comment (dannf) [inline ubuntu-server-netboot.py]: Can you drop this empty line?

File: autoinstall/README.md [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+     - as is `quiet splash root=/dev/ram0 ramdisk_size=1500000 ip=dhcp url=http://<server-provides-image>/focal-live-server-arm64.iso ---`
+     - to be `quiet splash root=/dev/ram0 ramdisk_size=1500000 ip=dhcp url=http://<server-provides-image>/focal-live-server-arm64.iso autoinstall "ds=nocloud-net;s=http://<server-provides-user-data>/" ---`
+
+## Trouble-shooting and Tips
+- The format of the first line "shebang", a.k.a. `#cloud-config` matters. Do not change it or input extra spaces
+  unless you know what you are doing.
+- The double quotes matter when injecting `user-data` in the kernel parameters.
+- The file name `user-data` is a must. Its the file name that `cloud-init` will look for.
+- The file `meta-data` is also a must, even it is an empty file in our example.
+- The trailing of the url serving `user-data` matters in the current autoinstall version.
```
Comment (dannf) [inline autoinstall/README.md]: I don't understand the `trailing of the url` - can you clarify?

**Review by dannf — CHANGES_REQUESTED** [review]
Thanks @tai271828 ! Think this is very close :) Comments inline.

File: autoinstall/README.md (hunk 16 above)
Comment (tai271828) [reply autoinstall/README.md]: Good catch. I meant "the trailing *slash*". That is to say, `http://<IP>/` vs. `http://<IP>`.

Additionally, for the reference, the format is defined in [NoCloud format of cloud-init](https://cloudinit.readthedocs.io/en/latest/topics/datasources/nocloud.html#datasource-nocloud). The document does not explicitly say the trailing slash is a must, and only demonstrates it in the example. I barely remember, but I think I have tried the format without the trailing slash and it indeed won't work as what [this stackoverflow thread](https://askubuntu.com/questions/1292032/how-can-current-ubuntu-versions-be-installed-via-pxe-network-boot-and-an-automat) says (search for "(including a trailing slash").

File: README.md (hunk 11 above)
Comment (tai271828) [reply README.md:25]: +1. Will do in the upcoming PR update.

File: README.md (hunk 12 above)
Comment (tai271828) [reply README.md:29]: +1. Will do in the upcoming PR update.

File: ubuntu-server-netboot.py (hunk 14 above)
Comment (tai271828) [reply ubuntu-server-netboot.py:263]: Good catch! Will do in the upcoming PR update.

File: ubuntu-server-netboot.py (hunk 15 above)
Comment (tai271828) [reply ubuntu-server-netboot.py]: +1. Will do in the upcoming PR update.

File: autoinstall/README.md (hunk 13 above)
Comment (tai271828) [reply autoinstall/README.md]: +1. Will do in the upcoming PR update.

**Comment by tai271828** [comment]
Hi @dannf , thank you for the most recent review. I have updated the pull request according to your change requests.

**Review by dannf — APPROVED** [review]
